src/query_plan/operation.rs

```diff
@@ -0,0 +1,435 @@
+                    let mut selection_set = field_entry.selection_set.clone();
+                    selection_set.selections = merged_selections;
+                    field_entry.make_mut().selection_set = selection_set;
+                    // field_entry.selection_set.selections = merged_selections;
+                }
+            }
+            Selection::FragmentSpread(named_fragment) => {
+                if let Some(fragment) = fragments.get(&named_fragment.fragment_name) {
+                    let expanded_selection_set = normalize_selections(
+                        &fragment.selection_set.selections.to_owned(),
+                        fragments,
+                    );
+                    normalized = merge_selections(&normalized, &expanded_selection_set);
+                } else {
+                    // no fragment found
+                }
+            }
+            Selection::InlineFragment(inline_fragment) => {
+                let expanded_selection_set = normalize_selections(
+                    &inline_fragment.selection_set.selections.to_owned(),
+                    fragments,
+                );
+
+                if let NormalizedSelection::NormalizedInlineFragment(fragment_entry) = normalized
+                    .entry(inline_fragment.into())
+                    .or_insert(NormalizedSelection::NormalizedInlineFragment(Node::new(
+                        NormalizedInlineFragment {
+                            type_condition: inline_fragment.type_condition.clone(),
+                            directives: inline_fragment.directives.clone(),
+                            selection_set: NormalizedSelectionSet {
+                                ty: inline_fragment.selection_set.ty.clone(),
+                                selections: IndexMap::new(),
+                            },
+                        },
+                    )))
+                {
+                    let merged_selections = merge_selections(
+                        &fragment_entry.selection_set.selections,
+                        &expanded_selection_set,
+                    );
+                    let mut selection_set = fragment_entry.selection_set.clone();
+                    selection_set.selections = merged_selections;
+                    fragment_entry.make_mut().selection_set = selection_set;
+                }
+            }
+        }
+    }
+    normalized
+}
+
+fn merge_selections(
+    source: &NormalizedSelectionMap,
+    to_merge: &NormalizedSelectionMap,
+) -> NormalizedSelectionMap {
+    let mut merged_selections = source.clone();
+    for (key, selection) in to_merge {
+        if source.contains_key(key) {
+            match selection {
+                NormalizedSelection::NormalizedField(field_to_merge) => {
+                    if let Some(NormalizedSelection::NormalizedField(source_field)) =
+                        merged_selections.get_mut(key)
+                    {
+                        // todo skip deferred
```

Comment:
This is for a later PR, but for the `@defer` handling, the gist is that two different `@defer` applications should be viewed as different for the sake of computing keys.

The JS codebase does this by essentially making the key computation spit out a random UUID for a `@defer` directive application. This doesn't work on its own though, since then the "same" directive application would output two different UUIDs if the key were computed twice on it. The JS codebase instead computes the keys for fields/fragment spreads/inline fragments once, and then stores the key in that object (so really the JS codebase is considering different fields/fragment spreads/inline fragments and not different `@defer` applications).

In this case, we could store a random UUID in the `NormalizedSelectionKey` (or use a static counter to generate it). Then, whenever we clone a key or field/fragment spread/inline fragment, we'd have to choose whether it's really the "same" or not (and if not, we change the UUID).

Reply:
https://github.com/apollographql/atlas-docs/blob/fed-deep-dive/federation_deepdive.md#query-cleanup says:
> 2) make sure all `@defer` application has a (unique) label (which simplify allows the planner to have an easy way to refer to the applications).

The Router does the same for auth. It does it by modifying every defer directive application:
* If it already has a `label` argument, add an arbitrary prefix to avoid collisions (Router uses `_`)
* If it doesn’t have a `label` argument, give it a unique one (Router uses a simple counter and `u32::to_string`, no randomness)

(This scheme is reverted when processing response data.)